sdk/python/kubeflow/training/api/training_client.py

```diff
@@ -81,6 +89,143 @@ def __init__(
+            )
+        except Exception as e:
+            print(e)
+
+        if isinstance(model_provider_parameters, HuggingFaceModelParams):
+            mp = "hf"
+
+        if isinstance(dataset_provider_parameters, S3DatasetParams):
+            dp = "s3"
+        elif isinstance(dataset_provider_parameters, HfDatasetParams):
+            dp = "hf"
+
+        # create init container spec
+        init_container_spec = utils.get_container_spec(
+            name=constants.STORAGE_CONTAINER,
+            image=constants.STORAGE_CONTAINER_IMAGE,
+            args=[
+                "--model_provider",
+                mp,
+                "--model_provider_parameters",
+                json.dumps(model_provider_parameters.__dict__),
+                "--dataset_provider",
+                dp,
+                "--dataset_provider_parameters",
+                json.dumps(dataset_provider_parameters.__dict__),
+            ],
+            volume_mounts=[
+                models.V1VolumeMount(name=constants.TRAINER_PV, mount_path="/workspace")
+            ],
+        )
+
+        # create app container spec
+        container_spec = utils.get_container_spec(
+            name=constants.JOB_PARAMETERS[constants.PYTORCHJOB_KIND]["container"],
+            image=constants.TRAINER_TRANSFORMER_IMAGE,
+            args=["--train_parameters", json.dumps(train_parameters.__dict__)],
+            volume_mounts=[
+                models.V1VolumeMount(name=constants.TRAINER_PV, mount_path="/workspace")
+            ],
+            resources=models.V1ResourceRequirements(requests=requests, limits=limits),
+        )
+
+        # create worker pod spec
+        worker_pod_template_spec = utils.get_pod_template_spec(
+            job_kind=constants.PYTORCHJOB_KIND,
+            containers_spec=[container_spec],
+            volumes_spec=[
+                models.V1Volume(
+                    name=constants.TRAINER_PV,
+                    persistent_volume_claim=models.V1PersistentVolumeClaimVolumeSource(
+                        claim_name=constants.TRAINER_PVC_NAME
+                    ),
+                )
+            ],
+        )
+
+        # create master pod spec
+        master_pod_template_spec = utils.get_pod_template_spec(
+            job_kind=constants.PYTORCHJOB_KIND,
+            containers_spec=[init_container_spec, container_spec],
+            volumes_spec=[
+                models.V1Volume(
+                    name=constants.TRAINER_PV,
+                    persistent_volume_claim=models.V1PersistentVolumeClaimVolumeSource(
+                        claim_name=constants.TRAINER_PVC_NAME
+                    ),
+                )
+            ],
+        )
+
+        job = utils.get_pytorchjob_template(
+            name=name,
+            namespace=namespace,
+            master_pod_template_spec=master_pod_template_spec,
+            worker_pod_template_spec=worker_pod_template_spec,
+            num_worker_replicas=num_workers - 1,
```

Comment:
Why  `num_worker_replicas=num_workers - 1` ?
Isn't it true that when num_workers = 2, we need to create PyTorchJob with 1 master and 2 workers ?

Reply:
I think when user specifies num_workers = 2 then he wants in total 2 pods

Reply:
@johnugeorge @deepanker13 @tenzen-y @kuizhiqing So for torch ddp master is the worker with 0 index where we initialize the Torch Distributed backend as mentioned here: https://pytorch.org/docs/stable/elastic/run.html#environment-variables.
But this master pod is still executes part of training.

Is that true ?

Reply:
I think `num_worker_replicas=num_workers`. And then, we should create one `Master` role replica regardless of the `num_workers`. Or, we may be able to create PyTorchJob without the `Master` role replica since we handle the Worker role with index 0 as a `Master` role.

Reply:
> How we can do it if the PyTorch initContainer will be only on pod which is created from .spec.Master replica ?

@andreyvelich IIUC, if there isn't `.replicaSpec[Master]`, the operator handles the Worker with index=0 as a `Master` replica.
So, couldn't we create PyTorchJob without the `Master` role?

Reply:
@tenzen-y So as you can see here: https://github.com/kubeflow/training-operator/blob/master/pkg/controller.v1/pytorch/initcontainer.go#L116-L120, we only set initContainer if `.replicaSpec[Master]` is set in the PyTorchJob spec.

Reply:
Also, looking to the PyTorch docs: https://pytorch.org/tutorials/intermediate/dist_tuto.html#initialization-methods, Master is the process with Rank 0, so Rank 0 process is coordinator and worker executor for rank 0.
So if user runs: `torchrun --nnodes=3 --nproc-per-node=1` or `create_job(num_workers=3)`, we should create PyTorchJob with the following spec:
```yaml
apiVersion: "kubeflow.org/v1"
kind: PyTorchJob
spec:
  pytorchReplicaSpecs:
    Master:
      replicas: 1
      ....
    Worker:
       replicas: 2
 nprocPerNode: 2 # if torchrun command is used
```

Like @kuizhiqing submitted in this example: https://github.com/kubeflow/training-operator/blob/master/examples/pytorch/cpu-demo/demo.yaml

Any thoughts @tenzen-y @kuizhiqing @johnugeorge @deepanker13 ?

Reply:
Yes, I agree with @andreyvelich , that's why I did num_worker_replicas=num_workers - 1

Reply:
> So as you can see here: https://github.com/kubeflow/training-operator/blob/master/pkg/controller.v1/pytorch/initcontainer.go#L116-L120, we only set initContainer if .replicaSpec[Master] is set in the PyTorchJob spec.

The `initContainer` function is for whether or not we set the `init-pytorch` intContainer, right?
So, I think that we can mutate the `storage-container` initContainer to any replicas including the `Worker` replicas via SDK.

However, I'm not strongly objections to adding the `storage-initializer` to the `Master` replica.
As I mentioned above, the solution is just one of the solutions.

Reply:
>So, I think that we can mutate the storage-container initContainer to any replicas including the Worker replicas via SDK.

Our idea was to only download model on the Master and then distribute it via shared PVC with `ReadOnlyMany` setting.

Reply:
> Our idea was to only download model and dataset on the Master and then distribute it via shared PVC with ReadOnlyMany setting.

That's right. Let's create a `Master` replica with the `storage-initializer`. Thank you for the clarifications :)